From 20e322931f8d976a450735224314650bb198c5c6 Mon Sep 17 00:00:00 2001
From: Aditya Kumar Singh <quic_adisi@quicinc.com>
Date: Wed, 14 Nov 2024 15:10:36 +0530
Subject: [PATCH 5/10] wifi: ath12k: add logic to select primary link

Current logic of primary selection uses debugfs as the only way to select
primary link. If debugfs is not provided, it will go with the assoc link
as primary link. In order to make primary link selection logic more robust,
following changes are being made -

1. If debugfs entry is there, give preference to it.
2. If it is not there, then use preferable hw_param enabled links as
   candidate links.
3. If no preferable hw_param enabled link exist, then consider whatever
   links are present as candidate links.
4. Now among this candidate links, select first set bit.

TODO: Enhance logic to use RSSI or some other metric to decide the
      primary link.

Signed-off-by: Aditya Kumar Singh <quic_adisi@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 22 ++++++++++++++-----
 1 file changed, 33 insertions(+) 13 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -21288,6 +21288,7 @@ static u8 ath12k_mac_ahsta_get_pri_link_
 	struct ath12k *ar;
 	u8 link_id = 0, pri_link_id;
 	bool is_link_found = false;
+	u16 pref_valid_links = 0, links_map;
 
 	lockdep_assert_held(&ah->conf_mutex);
 
@@ -21326,29 +21327,37 @@ static u8 ath12k_mac_ahsta_get_pri_link_
 	arvif = ath12k_get_arvif_from_link_id(ahvif, ahvif->primary_link_id);
 	if (arvif->ar->ab->hw_params->is_plink_preferable) {
 		pri_link_id = ahvif->primary_link_id;
-		return pri_link_id;
+		goto exit_pri_link_selection;
 	}
 
 select_pri_link:
-	/* if the configured link id is not present, then take assoc link
-	 * as the primary link
-	 */
-	pri_link_id = ahsta->assoc_link_id;
+	/* among all available links, get the preferable links bitmap */
+	for_each_set_bit(link_id, &valid_links, IEEE80211_MLD_MAX_NUM_LINKS) {
+		arvif = ath12k_get_arvif_from_link_id(ahvif, link_id);
+		if (!arvif || !arvif->ar)
+			continue;
 
-	arvif = ath12k_get_arvif_from_link_id(ahvif, ahsta->assoc_link_id);
-	if (arvif->ar->ab->hw_params->is_plink_preferable)
-		return pri_link_id;
+		if (!arvif->ar->ab->hw_params->is_plink_preferable)
+			continue;
 
-	/* if assoc link is not preferable, then select any other link as
-	 * primary link
-	 */
-	for_each_set_bit(link_id, &valid_links, IEEE80211_MLD_MAX_NUM_LINKS) {
-		if (link_id != ahsta->primary_link_id) {
-			pri_link_id = link_id;
-			break;
-		}
+		pref_valid_links |= BIT(link_id);
 	}
 
+	/* all the links available currently are not preferable but no
+	 * other choice hence need to select among these
+	 */
+	if (!pref_valid_links)
+		links_map = valid_links;
+	else
+		links_map = pref_valid_links;
+
+	/* TODO: Currently just selecting using ffs(). Proper logic can be
+	 * 	used here to select among these by using some other run
+	 *	time parameters like RSSI.
+	 */
+	pri_link_id = ffs(links_map) - 1;
+
+exit_pri_link_selection:
 	ath12k_mac_assign_middle_link_id(sta, ahsta, &pri_link_id);
 
 	return pri_link_id;
